src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionCodeFixProvider.Rewriter.cs

```diff
@@ -0,0 +1,106 @@
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertSwitchStatementToExpression
+{
+    using static SyntaxFactory;
+
+    internal sealed partial class ConvertSwitchStatementToExpressionCodeFixProvider
+    {
+        private sealed class Rewriter : CSharpSyntaxVisitor<SyntaxNode>
+        {
+            private List<ExpressionSyntax> _assignmentTargetsOpt;
+
+            private Rewriter()
+            {
+            }
+
+            public static ExpressionSyntax Rewrite(SwitchStatementSyntax node, out List<ExpressionSyntax> assignmentTargetsOpt)
+            {
+                var rewriter = new Rewriter();
+                var result = rewriter.VisitSwitchStatement(node);
+                assignmentTargetsOpt = rewriter._assignmentTargetsOpt;
+                return (ExpressionSyntax)result;
+            }
+
+            public override SyntaxNode VisitCasePatternSwitchLabel(CasePatternSwitchLabelSyntax node)
+            {
+                return node.Pattern;
+            }
+
+            public override SyntaxNode VisitCaseSwitchLabel(CaseSwitchLabelSyntax node)
+            {
+                return ConstantPattern(node.Value);
+            }
+
+            public override SyntaxNode VisitDefaultSwitchLabel(DefaultSwitchLabelSyntax node)
+            {
+                return DiscardPattern();
+            }
+
+            public override SyntaxNode VisitSwitchSection(SwitchSectionSyntax node)
+            {
+                Debug.Assert(node.Labels.Count == 1);
+                var switchLabel = node.Labels[0];
+                return SwitchExpressionArm(
+                    pattern: (PatternSyntax)Visit(switchLabel),
+                    whenClause: switchLabel.IsKind(SyntaxKind.CasePatternSwitchLabel, out CasePatternSwitchLabelSyntax patternSwitchLabel)
+                        ? patternSwitchLabel.WhenClause
+                        : null,
+                    expression: VisitStatements(node.Statements));
+            }
+
+            public override SyntaxNode VisitAssignmentExpression(AssignmentExpressionSyntax node)
+            {
+                (_assignmentTargetsOpt ?? (_assignmentTargetsOpt = new List<ExpressionSyntax>())).Add(node.Left);
+                return node.Right;
+            }
+
+            private ExpressionSyntax VisitStatements(SyntaxList<StatementSyntax> statements)
+            {
+                Debug.Assert(statements.Count > 0);
+                Debug.Assert(!statements[0].IsKind(SyntaxKind.BreakStatement));
+
+                var nodes = new List<ExpressionSyntax>();
+                foreach (var statement in statements)
+                {
+                    if (statement.IsKind(SyntaxKind.BreakStatement))
+                    {
+                        break;
+                    }
+
+                    nodes.Add((ExpressionSyntax)Visit(statement));
```

Comment:
you're arbitrarily recursing on statements here.  which makes me feel like any initial invariants you checked on the switch may not hold for nested switches.  Please add a bunch of tests for these scenarios.

Reply:
analyzer recurses on any switch, so the result is recursively combined and we can be sure about nested switches as well. also, any node that is not handled (doesn't have an override) will fallback to default visit which will signify a failure.

so in the rewriter, default visit is practically unreachable.

def needs more tests.